### DISCUSSION TITLE
How to create components in commonMain for a Compose Multiplatform Project?

### DISCUSSION THREAD

**Discussion by ksharma-xyz** [evant/kotlin-inject#455] [Q&A]
I have been stuck trying to use kotlin-inject library therefore asking for help and I really wanna use this over koin because of compile time resolving. Therefore appreciate any help / directions and also I might just be asking some stupid questions which are covered in docs but I'm not able to get my head around it, so explaining my use case for better understanding. 

Version:
ksp = "2.0.21-1.0.27"
kotlin = "2.0.21"
kotlinInject = "0.7.2"

**Use case**
:network -> using ktor to define http client, need to create a singleton HttpClient object scoped to the Application lifecycle.
:database -> has interface and internal implementation, will store data in db. Need to scope this to Application Lifecycle.
:ui  -> Has ViewModels and calls Repository from network module. Need to scope ViewModel to the scope of Screen level composable.

All the code is in `commonMain` and none of it has `iosMain` / `andoridMain` implementation as of now - **Will this be a problem ?**

`NetworkComponent`

```kotlin
@Component
abstract class NetworkComponent {

    val RateLimiterImpl.bind: RateLimiter
        @Provides get() = this
    
    @Provides
    fun provideHttpClient(): HttpClient = getHttpClient()
}

fun getHttpClient(): HttpClient {
    return HttpClient {
        expectSuccess = true
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                isLenient = true
                prettyPrint = true
            })
        }
    }
}

```

for ViewModel I have something like this:

```kotlin
@Component
abstract class ViewModelComponent {

    @Provides
    fun provideABCModel(
        db: MyDatabase,
    ): ABCViewModel = ABCViewModel(db)

    companion object
}
```

Problems I'm facing:

1. `create` method is not getting generated for these components e.g. `ViewModelComponent::class.create()`

**Comment by evant** [comment]
You check out https://github.com/evant/kotlin-inject/blob/main/docs/multiplatform.md?

**Reply by ksharma-xyz** [reply]
Yea, I did, and tried everything, but seems like `create()` method is not resolved. Not sure what is breaking up.

**Reply by ksharma-xyz** [reply]
I gave it another go, understood that create methods will not be created for `commonMain`.

Although now am stuck at another place, even though I'm providing the `SqlDriver`, the compilation fails. 
I'm providing the `SqlDriver` in Android / ios source sets like below:

## ComonMain Source Set:

**SandookDriverFactory.kt**
```
@Inject
class SandookDriverFactory(private val driver: SqlDriver) {
    fun build(): KrailSandook = KrailSandook(driver = driver)
}
```

DI :

**SandookComponent.kt**
```kotlin
import me.tatarka.inject.annotations.Component
import me.tatarka.inject.annotations.KmpComponentCreate
import me.tatarka.inject.annotations.Provides

@Component
abstract class SandookComponent {

    internal val RealSandookDb.bind: SandookDb
        @Provides get() = this

    abstract val sandookDb: SandookDb

    companion object
}

@KmpComponentCreate
expect fun createSandookComponent(): SandookComponent
```

**SandookFactory.kt**
```kotlin
@Inject
class SandookFactory(private val driver: SqlDriver) {
    fun build(): KrailSandook = KrailSandook(driver = driver)
}
```

**SQLComponent.kt**

```kotlin
expect interface SQLPlatformComponent

@Component
interface SQLComponent : SQLPlatformComponent {

    @Provides
    fun provideSqlDelightDatabase(
    factory: SandookFactory,
    ): KrailSandook = factory.build()
}
```

## AndroidMain source set

```kotlin
@Component
actual interface SQLPlatformComponent {

    @Provides
    fun provideSQLDriver(application: Application): SqlDriver =
        AndroidSqliteDriver(
            schema = KrailSandook.Schema,
            context = application,
            name = "krailSandook.db"
        )

    companion object
}
```

## iosMain source set

```kotlin
@Component
actual interface SQLPlatformComponent {

    @Provides
    fun provideSQLDriver(): SqlDriver =
        NativeSqliteDriver(schema = KrailSandook.Schema, name = "krailSandook.db")

    companion object
}
```

Error:
`[ksp] Cannot find an @Inject constructor or provider for: app.cash.sqldelight.db.SqlDriver`
